PhantomJS doesn't shut down properly on Linux

Hi, [details are here](https://groups.google.com/forum/#!topic/scala-js/Y6k10K5YyKI), basically what happens is that utest suite fails and PhantomJS OS process still exists which results in unpredictable errors when you run test suite again. This is the first run that ends with error : 
```
[info] Compiling 1 Scala source to /.../js/target/scala-2.11/test-classes...
[info] Fast optimizing /.../js/target/scala-2.11/app-test-fastopt.js
Uncaught failure in async execution: utest.AssertionError: assert(p == expectedPosition)

  /tmp/phantomjs-launcher4242163106310882054.js:9 in onError


  /tmp/phantomjs-launcher4242163106310882054.js:11 in onError
  file:///.../js/target/scala-2.11/app-test-fastopt.js:94303 (in function "doWriteLine__p4__T__V")

  /tmp/phantomjs-launcher4242163106310882054.js:13
[trace] Stack trace suppressed: run last js/test:test for the full output.
[error] Could not run test ....EditorSuite: org.scalajs.jsenv.ComJSEnv$ComClosedException: JSCom has been closed
[trace] Stack trace suppressed: run last js/test:executeTests for the full output.
[error] (js/test:executeTests) org.scalajs.jsenv.ComJSEnv$ComClosedException: Phantom.js isn't connected
```
Then you have to go and kill the process for the next run to work correctly : 
```
ps aux | grep phantom
user     7310  0.2  0.7 1739472 61520 pts/1   Sl+  12:50   0:00 phantomjs /tmp/phantomjs-launcher7268336573458704946.js
kill -9 7310
```

Otherwise your suite probably fails on some IndexedDB issues or something that doesn't allow for running multiple PhantomJS instances...

When I'm chasing test issues I'm constantly running suite and killing PhantomJS afterwards. 

The solution would be checking whether PhantomJS process isn't already running before starting tests. It seems to be much easier that checking whether the process shut down properly.

NOTE: It might be specific to my test suite, but I didn't figure out the reason, I'm closing all resources I'm using when error occurs, like IndexedDb. Dunno why the process hangs out.

Basically : 
```scala
    "test"-{
      try {
        Observable.fromIterable(List(1,2,3)).unsafeSubscribe(
          new Observer[Int] {
            override def onError(ex: Throwable): Unit = throw new IllegalStateException("wtf")
            override def onComplete(): Unit = throw new IllegalStateException("wtf")
            override def onNext(elem: Int): Future[Ack] = Continue
          }
        )
      } catch {
        case e : Throwable => println("NEVER PRINTS")
      }
      println("NEVER PRINTS")
    }
```

results in this ouput, which leaves PhantomJS process running : 
```
[info] Compiling 1 Scala source to /.../js/target/scala-2.11/test-classes...
[info] Fast optimizing /.../js/target/scala-2.11/app-test-fastopt.js
Uncaught failure in async execution: utest.AssertionError: assert(p == expectedPosition)

  /tmp/phantomjs-launcher4242163106310882054.js:9 in onError


  /tmp/phantomjs-launcher4242163106310882054.js:11 in onError
  file:///.../js/target/scala-2.11/app-test-fastopt.js:94303 (in function "doWriteLine__p4__T__V")

  /tmp/phantomjs-launcher4242163106310882054.js:13
[trace] Stack trace suppressed: run last js/test:test for the full output.
[error] Could not run test ....EditorSuite: org.scalajs.jsenv.ComJSEnv$ComClosedException: JSCom has been closed
[trace] Stack trace suppressed: run last js/test:executeTests for the full output.
[error] (js/test:executeTests) org.scalajs.jsenv.ComJSEnv$ComClosedException: Phantom.js isn't connected
```

--- a/js-envs/src/main/scala/org/scalajs/jsenv/AsyncJSRunner.scala
+++ b/js-envs/src/main/scala/org/scalajs/jsenv/AsyncJSRunner.scala
@@ -15,7 +15,17 @@ trait AsyncJSRunner {
    */
   def start(): Future[Unit]
 
-  /** Abort the associated run */
+  /** Aborts the associated run.
+   *
+   *  There is no guarantee that the runner will be effectively terminated
+   *  by the time this method returns. If necessary, this call can be followed
+   *  by a call to `await()`.
+   *
+   *  If the run has already completed, this does nothing. Similarly,
+   *  subsequent calls to `stop()` will do nothing.
+   *
+   *  This method cannot be called before `start()` has been called.
+   */
   def stop(): Unit
 
   /**

--- a/js-envs/src/main/scala/org/scalajs/jsenv/AsyncJSRunner.scala
+++ b/js-envs/src/main/scala/org/scalajs/jsenv/AsyncJSRunner.scala
@@ -48,12 +48,31 @@ trait AsyncJSRunner {
 
   /** Await completion of the started Run for the duration specified
    *  by [[atMost]]. Strictly equivalent to:
-   * 
+   *
    *  {{{
    *  Await.result(future, atMost)
    *  }}}
-   * 
+   *
    */
   final def await(atMost: Duration): Unit = Await.result(future, atMost)
+
+  /** Awaits completion of the started Run for the duration specified by
+   *  [[atMost]], or force it to stop.
+   *
+   *  If any exception is thrown while awaiting completion (including a
+   *  [[scala.concurrent.TimeoutException TimeoutException]], forces the runner
+   *  to stop by calling `stop()` before rethrowing the exception.
+   *
+   *  Strictly equivalent to:
+   *
+   *  {{{
+   *  try await(atMost)
+   *  finally stop()
+   *  }}}
+   */
+  final def awaitOrStop(atMost: Duration): Unit = {
+    try await(atMost)
+    finally stop()
+  }
 
 }

--- a/js-envs/src/main/scala/org/scalajs/jsenv/ComJSRunner.scala
+++ b/js-envs/src/main/scala/org/scalajs/jsenv/ComJSRunner.scala
@@ -1,14 +1,27 @@
 package org.scalajs.jsenv
 
+import scala.concurrent.duration.Duration
+
 trait ComJSRunner extends AsyncJSRunner {
 
   /** Send a message to the JS VM. Throws if the message cannot be sent. */
   def send(msg: String): Unit
 
-  /** Block until a message is received. Throws a [[ComClosedExcpetion]]
-   *  if the channel is closed before a message is received.
+  /** Blocks until a message is received and returns it.
+   *
+   *  @throws ComClosedException if the channel is closed before a message
+   *    is received
    */
-  def receive(): String
+  final def receive(): String = receive(Duration.Inf)
+
+  /** Blocks until a message is received and returns it.
+   *
+   *  @throws ComClosedException if the channel is closed before a message
+   *    is received
+   *  @throws scala.concurrent.TimeoutException if the timeout expires
+   *    before a message is received and the channel is still open
+   */
+  def receive(timeout: Duration): String
 
   /** Close the communication channel. Allows the VM to terminate if it is
    *  still waiting for callback. The JVM side **must** call close in

--- a/js-envs/src/main/scala/org/scalajs/jsenv/RetryingComJSEnv.scala
+++ b/js-envs/src/main/scala/org/scalajs/jsenv/RetryingComJSEnv.scala
@@ -14,6 +14,7 @@ import org.scalajs.core.tools.classpath._
 import org.scalajs.core.tools.logging._
 
 import scala.concurrent.{Future, Promise, ExecutionContext}
+import scala.concurrent.duration.Duration
 import scala.collection.mutable
 import scala.annotation.tailrec
 import scala.util.control.NonFatal
@@ -21,7 +22,7 @@ import scala.util.{Try, Failure, Success}
 
 /** A RetryingComJSEnv allows to automatically retry if a call to the underlying
  *  ComJSRunner fails.
- *  
+ *
  *  While it protects the JVM side from observing state that differs inbetween
  *  runs that have been retried, it assumes that the executed JavaScript code
  *  does not have side-effects other than the ones visible through the channel
@@ -57,7 +58,7 @@ final class RetryingComJSEnv(baseEnv: ComJSEnv,
 
   private class RetryingComJSRunner(classpath: CompleteClasspath,
       code: VirtualJSFile, logger: Logger,
-      console: JSConsole) extends DummyJSRunner with ComJSRunner { 
+      console: JSConsole) extends DummyJSRunner with ComJSRunner {
 
     private[this] val promise = Promise[Unit]
 
@@ -88,12 +89,12 @@ final class RetryingComJSEnv(baseEnv: ComJSEnv,
       logAndDo(Send(msg))
     }
 
-    def receive(): String = {
+    def receive(timeout: Duration): String = {
       @tailrec
       def recLoop(): String = {
         // Need to use Try for tailrec
         Try {
-          val result = curRunner.receive()
+          val result = curRunner.receive(timeout)
           // At this point, we are sending state to the JVM, we cannot retry
           // after this.
           hasReceived = true

--- a/js-envs/src/main/scala/org/scalajs/jsenv/Utils.scala
+++ b/js-envs/src/main/scala/org/scalajs/jsenv/Utils.scala
@@ -1,0 +1,34 @@
+/*                     __                                               *\
+**     ________ ___   / /  ___      __ ____  Scala.js JS environments   **
+**    / __/ __// _ | / /  / _ | __ / // __/  (c) 2013, LAMP/EPFL        **
+**  __\ \/ /__/ __ |/ /__/ __ |/_// /_\ \    http://scala-js.org/       **
+** /____/\___/_/ |_/____/_/ | |__/ /____/                               **
+**                          |/____/                                     **
+\*                                                                      */
+
+
+package org.scalajs.jsenv
+
+import scala.concurrent.duration._
+
+private[jsenv] object Utils {
+  final class OptDeadline private (
+      val deadline: Deadline /* nullable */) extends AnyVal {
+    def millisLeft: Long =
+      if (deadline == null) 0
+      else (deadline.timeLeft.toMillis max 1L)
+
+    def isOverdue: Boolean =
+      if (deadline == null) false
+      else deadline.isOverdue
+  }
+
+  object OptDeadline {
+    def apply(timeout: Duration): OptDeadline = {
+      new OptDeadline(timeout match {
+        case timeout: FiniteDuration => timeout.fromNow
+        case _                       => null
+      })
+    }
+  }
+}

--- a/js-envs/src/main/scala/org/scalajs/jsenv/nodejs/NodeJSEnv.scala
+++ b/js-envs/src/main/scala/org/scalajs/jsenv/nodejs/NodeJSEnv.scala
@@ -10,6 +10,7 @@
 package org.scalajs.jsenv.nodejs
 
 import org.scalajs.jsenv._
+import org.scalajs.jsenv.Utils.OptDeadline
 
 import org.scalajs.core.ir.Utils.escapeJS
 
@@ -21,6 +22,8 @@ import org.scalajs.core.tools.logging._
 import java.io.{ Console => _, _ }
 import java.net._
 
+import scala.concurrent.TimeoutException
+import scala.concurrent.duration._
 import scala.io.Source
 
 class NodeJSEnv private (
@@ -174,16 +177,32 @@ class NodeJSEnv private (
       }
     }
 
-    def receive(): String = {
+    def receive(timeout: Duration): String = {
       if (!awaitConnection())
         throw new ComJSEnv.ComClosedException("Node.js isn't connected")
+
+      js2jvm.mark(Int.MaxValue)
+      val savedSoTimeout = comSocket.getSoTimeout()
       try {
+        val optDeadline = OptDeadline(timeout)
+
+        comSocket.setSoTimeout((optDeadline.millisLeft min Int.MaxValue).toInt)
         val len = js2jvm.readInt()
-        val carr = Array.fill(len)(js2jvm.readChar())
+        val carr = Array.fill(len) {
+          comSocket.setSoTimeout((optDeadline.millisLeft min Int.MaxValue).toInt)
+          js2jvm.readChar()
+        }
+
+        js2jvm.mark(0)
         String.valueOf(carr)
       } catch {
         case e: EOFException =>
           throw new ComJSEnv.ComClosedException(e)
+        case e: SocketTimeoutException =>
+          js2jvm.reset()
+          throw new TimeoutException("Timeout expired")
+      } finally {
+        comSocket.setSoTimeout(savedSoTimeout)
       }
     }
 

--- a/js-envs/src/main/scala/org/scalajs/jsenv/phantomjs/PhantomJSEnv.scala
+++ b/js-envs/src/main/scala/org/scalajs/jsenv/phantomjs/PhantomJSEnv.scala
@@ -10,6 +10,7 @@
 package org.scalajs.jsenv.phantomjs
 
 import org.scalajs.jsenv._
+import org.scalajs.jsenv.Utils.OptDeadline
 
 import org.scalajs.core.ir.Utils.{escapeJS, fixFileURI}
 
@@ -25,6 +26,7 @@ import scala.collection.mutable
 import scala.annotation.tailrec
 
 import scala.concurrent.{ExecutionContext, TimeoutException}
+import scala.concurrent.duration.Duration
 
 class PhantomJSEnv(
     phantomjsPath: String = "phantomjs",
@@ -107,6 +109,7 @@ class PhantomJSEnv(
     def log(msg: String): Unit = logger.debug(s"PhantomJS WS Jetty: $msg")
 
     private[this] val recvBuf = mutable.Queue.empty[String]
+    private[this] val fragmentsBuf = new StringBuilder
 
     mgr.start()
 
@@ -241,34 +244,52 @@ class PhantomJSEnv(
       }
     }
 
-    def receive(): String = synchronized {
+    def receive(timeout: Duration): String = synchronized {
       if (recvBuf.isEmpty && !awaitConnection())
         throw new ComJSEnv.ComClosedException("Phantom.js isn't connected")
 
-      @tailrec
-      def loop(acc: String): String = {
-        val frag = receiveFrag()
-        val newAcc = acc + frag.substring(1)
+      val deadline = OptDeadline(timeout)
 
-        if (frag(0) == '0')
-          newAcc
-        else if (frag(0) == '1')
-          loop(newAcc)
-        else
+      @tailrec
+      def loop(): String = {
+        /* The fragments are accumulated in an instance-wide buffer in case
+         * receiving a non-first fragment times out.
+         */
+        val frag = receiveFrag(deadline)
+        fragmentsBuf ++= frag.substring(1)
+
+        if (frag(0) == '0') {
+          val result = fragmentsBuf.result()
+          fragmentsBuf.clear()
+          result
+        } else if (frag(0) == '1') {
+          loop()
+        } else {
           throw new AssertionError("Bad fragmentation flag in " + frag)
+        }
       }
 
-      loop("")
+      try {
+        loop()
+      } catch {
+        case e: Throwable if !e.isInstanceOf[TimeoutException] =>
+          fragmentsBuf.clear() // the protocol is broken, so discard the buffer
+          throw e
+      }
     }
 
-    private def receiveFrag(): String = {
-      while (recvBuf.isEmpty && !mgr.isClosed)
-        wait()
+    private def receiveFrag(deadline: OptDeadline): String = {
+      while (recvBuf.isEmpty && !mgr.isClosed && !deadline.isOverdue)
+        wait(deadline.millisLeft)
 
-      if (recvBuf.isEmpty)
-        throw new ComJSEnv.ComClosedException
-      else
-        recvBuf.dequeue()
+      if (recvBuf.isEmpty) {
+        if (mgr.isClosed)
+          throw new ComJSEnv.ComClosedException
+        else
+          throw new TimeoutException("Timeout expired")
+      }
+
+      recvBuf.dequeue()
     }
 
     def close(): Unit = mgr.stop()

--- a/js-envs/src/main/scala/org/scalajs/jsenv/rhino/RhinoJSEnv.scala
+++ b/js-envs/src/main/scala/org/scalajs/jsenv/rhino/RhinoJSEnv.scala
@@ -10,6 +10,7 @@
 package org.scalajs.jsenv.rhino
 
 import org.scalajs.jsenv._
+import org.scalajs.jsenv.Utils.OptDeadline
 
 import org.scalajs.core.tools.sem.Semantics
 import org.scalajs.core.tools.io._
@@ -22,7 +23,7 @@ import scala.io.Source
 
 import scala.collection.mutable
 
-import scala.concurrent.{Future, Promise, Await}
+import scala.concurrent.{Future, Promise, Await, TimeoutException}
 import scala.concurrent.duration._
 
 import scala.reflect.ClassTag
@@ -108,9 +109,9 @@ final class RhinoJSEnv private (
 
     def send(msg: String): Unit = channel.sendToJS(msg)
 
-    def receive(): String = {
+    def receive(timeout: Duration): String = {
       try {
-        channel.recvJVM()
+        channel.recvJVM(timeout)
       } catch {
         case _: ChannelClosedException =>
           throw new ComJSEnv.ComClosedException
@@ -443,19 +444,23 @@ object RhinoJSEnv {
     def sendToJS(msg: String): Unit = synchronized {
       ensureOpen(_closedJVM)
       jvm2js.enqueue(msg)
-      notify()
+      notifyAll()
     }
 
     def sendToJVM(msg: String): Unit = synchronized {
       ensureOpen(_closedJS)
       js2jvm.enqueue(msg)
-      notify()
+      notifyAll()
     }
 
-    def recvJVM(): String = synchronized {
-      while (js2jvm.isEmpty && ensureOpen(_closedJS))
-        wait()
+    def recvJVM(timeout: Duration): String = synchronized {
+      val deadline = OptDeadline(timeout)
 
+      while (js2jvm.isEmpty && ensureOpen(_closedJS) && !deadline.isOverdue)
+        wait(deadline.millisLeft)
+
+      if (js2jvm.isEmpty)
+        throw new TimeoutException("Timeout expired")
       js2jvm.dequeue()
     }
 
@@ -482,12 +487,12 @@ object RhinoJSEnv {
 
     def closeJS(): Unit = synchronized {
       _closedJS = true
-      notify()
+      notifyAll()
     }
 
     def closeJVM(): Unit = synchronized {
       _closedJVM = true
-      notify()
+      notifyAll()
     }
 
     /** Throws if the channel is closed and returns true */

--- a/js-envs/src/test/scala/org/scalajs/jsenv/test/AsyncTests.scala
+++ b/js-envs/src/test/scala/org/scalajs/jsenv/test/AsyncTests.scala
@@ -12,7 +12,7 @@ import org.junit.Assert._
 import scala.concurrent.Await
 import scala.concurrent.duration._
 
-/** A couple of tests that test communication for mix-in into a test suite */
+/** A couple of tests that test async runners for mix-in into a test suite */
 trait AsyncTests {
 
   protected final val DefaultTimeout: Duration = 10.seconds
@@ -35,6 +35,16 @@ trait AsyncTests {
     Await.result(fut, DefaultTimeout)
 
     assertFalse("VM should be terminated", runner.isRunning)
+  }
+
+  @Test
+  def stopAfterTerminatedTest = {
+    val runner = asyncRunner("")
+    val fut = runner.start()
+
+    Await.result(fut, DefaultTimeout)
+
+    runner.stop() // should do nothing, and not fail
   }
 
 }

--- a/js-envs/src/test/scala/org/scalajs/jsenv/test/ComTests.scala
+++ b/js-envs/src/test/scala/org/scalajs/jsenv/test/ComTests.scala
@@ -53,6 +53,8 @@ trait ComTests extends AsyncTests {
 
     com.close()
     com.await(DefaultTimeout)
+
+    com.stop() // should do nothing, and not fail
   }
 
   def comCloseJSTestCommon(timeout: Long) = {
@@ -75,6 +77,8 @@ trait ComTests extends AsyncTests {
 
     com.close()
     com.await(DefaultTimeout)
+
+    com.stop() // should do nothing, and not fail
   }
 
   @Test

--- a/js-envs/src/test/scala/org/scalajs/jsenv/test/RetryingComJSEnvTest.scala
+++ b/js-envs/src/test/scala/org/scalajs/jsenv/test/RetryingComJSEnvTest.scala
@@ -9,6 +9,7 @@ import org.scalajs.core.tools.logging._
 import org.scalajs.core.tools.sem._
 
 import scala.concurrent.Future
+import scala.concurrent.duration.Duration
 
 import org.junit.Test
 
@@ -50,7 +51,7 @@ class RetryingComJSEnvTest extends JSEnvTest with ComTests {
 
     private class FailingComJSRunner(baseRunner: ComJSRunner)
         extends DummyJSRunner with ComJSRunner {
-      
+
       def future = baseRunner.future
 
       def send(msg: String): Unit = {
@@ -58,12 +59,12 @@ class RetryingComJSEnvTest extends JSEnvTest with ComTests {
         baseRunner.send(msg)
       }
 
-      def receive(): String = {
+      def receive(timeout: Duration): String = {
         if (shouldFail) {
           failedReceive = true
           fail()
         }
-        baseRunner.receive()
+        baseRunner.receive(timeout)
       }
 
 
@@ -85,7 +86,7 @@ class RetryingComJSEnvTest extends JSEnvTest with ComTests {
       private def shouldFail = !failedReceive && fails < maxFails
 
       private def maybeFail() = {
-        if (shouldFail) 
+        if (shouldFail)
           fail()
       }
 

--- a/js-envs/src/test/scala/org/scalajs/jsenv/test/RetryingComJSEnvTest.scala
+++ b/js-envs/src/test/scala/org/scalajs/jsenv/test/RetryingComJSEnvTest.scala
@@ -67,7 +67,6 @@ class RetryingComJSEnvTest extends JSEnvTest with ComTests {
         baseRunner.receive(timeout)
       }
 
-
       def start(): Future[Unit] = {
         maybeFail()
         baseRunner.start()

--- a/js-envs/src/test/scala/org/scalajs/jsenv/test/TimeoutComTests.scala
+++ b/js-envs/src/test/scala/org/scalajs/jsenv/test/TimeoutComTests.scala
@@ -3,6 +3,7 @@ package org.scalajs.jsenv.test
 import org.junit.Test
 import org.junit.Assert._
 
+import scala.concurrent.TimeoutException
 import scala.concurrent.duration._
 
 trait TimeoutComTests extends TimeoutTests with ComTests {
@@ -49,6 +50,33 @@ trait TimeoutComTests extends TimeoutTests with ComTests {
       com.send(s"Hello World: $i")
       assertEquals(s"Got: Hello World: $i", com.receive())
       assertTrue("Execution took too little time", deadline.isOverdue())
+    }
+
+    com.close()
+    com.await(DefaultTimeout)
+
+  }
+
+  @Test
+  def receiveTimeoutTest = {
+
+    val com = comRunner(s"""
+      scalajsCom.init(function(msg) {
+        setTimeout(scalajsCom.send, 2000, "Got: " + msg);
+      });
+    """)
+
+    com.start()
+
+    for (i <- 1 to 2) {
+      com.send(s"Hello World: $i")
+      try {
+        com.receive(900.millis)
+        fail("Expected TimeoutException to be thrown")
+      } catch {
+        case _: TimeoutException =>
+      }
+      assertEquals(s"Got: Hello World: $i", com.receive(3000.millis))
     }
 
     com.close()

--- a/js-envs/src/test/scala/org/scalajs/jsenv/test/TimeoutComTests.scala
+++ b/js-envs/src/test/scala/org/scalajs/jsenv/test/TimeoutComTests.scala
@@ -124,6 +124,28 @@ trait TimeoutComTests extends TimeoutTests with ComTests {
       case t: Throwable => // all is well
     }
 
+    async.stop() // should do nothing, and not fail
+
+  }
+
+  @Test
+  def doubleStopTest = {
+    val async = asyncRunner(s"""
+      setInterval(function() {}, 0);
+    """)
+
+    async.start()
+    async.stop()
+    async.stop() // should do nothing, and not fail
+
+    try {
+      async.await(DefaultTimeout)
+      fail("Expected await to fail")
+    } catch {
+      case t: Throwable => // all is well
+    }
+
+    async.stop() // should do nothing, and not fail
   }
 
 }

--- a/test-adapter/src/main/scala/org/scalajs/sbttestadapter/ComUtils.scala
+++ b/test-adapter/src/main/scala/org/scalajs/sbttestadapter/ComUtils.scala
@@ -5,22 +5,41 @@ import org.scalajs.jsenv._
 
 import scala.annotation.tailrec
 
+import scala.concurrent.duration._
+
 private[testadapter] object ComUtils {
 
   type Handler[+T] = PartialFunction[(String, String), T]
   type LoopHandler[+T] = Handler[Option[T]]
 
+  def receiveLoop[T](com: ComJSRunner)(handler: LoopHandler[T]): T =
+    receiveLoop(com, Duration.Inf)(handler)
+
   @tailrec
-  def receiveLoop[T](com: ComJSRunner)(handler: LoopHandler[T]): T = {
-    receiveResponse(com)(handler) match {
+  def receiveLoop[T](com: ComJSRunner, timeout: Duration)(
+      handler: LoopHandler[T]): T = {
+    receiveResponse(com, timeout)(handler) match {
       case Some(v) => v
-      case None => receiveLoop(com)(handler)
+      case None    => receiveLoop(com, timeout)(handler)
     }
   }
 
-  def receiveResponse[T](com: ComJSRunner)(handler: Handler[T]): T = {
+  @tailrec
+  def receiveLoop[T](com: ComJSRunner, deadline: Deadline)(
+      handler: LoopHandler[T]): T = {
+    receiveResponse(com, deadline.timeLeft)(handler) match {
+      case Some(v) => v
+      case None    => receiveLoop(com, deadline)(handler)
+    }
+  }
+
+  def receiveResponse[T](com: ComJSRunner)(handler: Handler[T]): T =
+    receiveResponse(com, Duration.Inf)(handler)
+
+  def receiveResponse[T](com: ComJSRunner, timeout: Duration)(
+      handler: Handler[T]): T = {
     val resp = {
-      try com.receive()
+      try com.receive(timeout)
       catch {
         case t: ComJSEnv.ComClosedException =>
           // Check if runner failed. If it did, throw that exception instead

--- a/test-adapter/src/main/scala/org/scalajs/sbttestadapter/ScalaJSRunner.scala
+++ b/test-adapter/src/main/scala/org/scalajs/sbttestadapter/ScalaJSRunner.scala
@@ -17,6 +17,9 @@ import org.scalajs.jsenv._
 
 import scala.collection.concurrent.TrieMap
 
+import scala.concurrent.duration._
+import scala.util.{Try, Failure, Success}
+
 import java.util.concurrent.atomic.AtomicInteger
 
 import sbt.testing._
@@ -64,17 +67,35 @@ final class ScalaJSRunner private[testadapter] (
   def done(): String = synchronized {
     ensureNotDone()
 
-    stopSlaves()
+    /* Whatever happens in here, we must close and eventually terminate all
+     * VMs. So we capture all exceptions in Try's, and we'll rethrow one of
+     * them (if any) at the end.
+     */
 
-    master.send("runnerDone")
-    val summary = ComUtils.receiveResponse(master) {
-      case ("ok", summary) => summary
+    // First we run the stopping sequence of the slaves
+    val slavesDeadline = VMTermTimeout.fromNow
+    val slavesClosing = stopSlaves(slavesDeadline)
+
+    /* Once all slaves are closing, we can schedule termination of the master.
+     * We need a fresh deadline for the master, since we can only start its
+     * scheduling when the slaves are closing.
+     * If we used the same deadline, and a slave timed out during its stopping
+     * sequence, the master would have 0 ms to stop, which is not fair.
+     */
+    val masterDeadline = VMTermTimeout.fromNow
+    val summaryTry = Try {
+      master.send("runnerDone")
+      val summary = ComUtils.receiveResponse(master, masterDeadline.timeLeft) {
+        case ("ok", summary) => summary
+      }
+      master.close()
+      summary
     }
-    master.close()
 
-    // Await all JS VMs
-    slaves.values.foreach(_.await(VMTermTimeout))
-    master.await(VMTermTimeout)
+    // Now we wait for everyone to be completely stopped
+    val slavesStopped =
+      slaves.values.toList.map(s => Try(s.awaitOrStop(slavesDeadline.timeLeft)))
+    val masterStopped = Try(master.awaitOrStop(masterDeadline.timeLeft))
 
     // Cleanup
     master = null
@@ -82,7 +103,13 @@ final class ScalaJSRunner private[testadapter] (
 
     framework.runDone()
 
-    summary
+    // At this point, rethrow any exception we captured on the way with Try's
+    slavesClosing.get
+    slavesStopped.foreach(_.get)
+    masterStopped.get
+
+    // And finally, if all went well, return the summary
+    summaryTry.get
   }
 
   // Runner Messaging
@@ -122,17 +149,34 @@ final class ScalaJSRunner private[testadapter] (
     }
   }
 
-  private def stopSlaves(): Unit = {
-    val slaves = this.slaves.values
+  /** Starts the stopping sequence of all slaves.
+   *  The returned future will be completed when all slaves are closing.
+   */
+  private def stopSlaves(deadline: Deadline): Try[Unit] = {
+    val slaves = this.slaves.values.toList // .toList to make it strict
 
-    slaves.foreach(_.send("stopSlave"))
-    slaves foreach { slave =>
-      ComUtils.receiveLoop(slave)(msgHandler(slave) orElse ComUtils.doneHandler)
+    // First launch the stopping sequence on all slaves
+    val stopMessagesSent = for (slave <- slaves) yield Try {
+      slave.send("stopSlave")
     }
-    slaves.foreach(_.close())
+
+    // Then process all their messages and close them
+    val slavesClosed = for (slave <- slaves) yield Try {
+      ComUtils.receiveLoop(slave, deadline)(
+          msgHandler(slave) orElse ComUtils.doneHandler)
+      slave.close()
+    }
+
+    // Return the first failed of all these Try's
+    (stopMessagesSent ++ slavesClosed) collectFirst {
+      case failure: Failure[Unit] => failure
+    } getOrElse Success(())
   }
 
   private def createSlave(): ComJSRunner = {
+    // We don't want to create new slaves when we're closing/closed
+    ensureNotDone()
+
     // Launch the slave
     val slave = framework.createRunner(slaveLauncher)
     slave.start()

